### PR TITLE
Add patterns to canonicalize ForOp and AffineMin based on dispatch ID range

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  %3 = flow.dispatch.workgroups[%c4, %c4, %c4] (%arg0) : (tensor<8x8xf32>)
+    -> tensor<8x8xf32> = (%arg1 : !flow.dispatch.input<8x8xf32>, %arg2 : !flow.dispatch.output<8x8xf32>) {
+    %c1 = constant 1 : index
+    %c2 = constant 2 : index
+    %c8 = constant 8 : index
+    %workgroup_id_2 = flow.dispatch.workgroup.id[2] : index
+    %workgroup_count_2 = flow.dispatch.workgroup.count[2] : index
+    %4 = affine.apply #map0(%workgroup_id_2)
+    %5 = affine.apply #map0(%workgroup_count_2)
+    scf.for %arg4 = %4 to %c8 step %5 {
```
Comment (benvanik) [inline iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: We should test this with dynamic shapes as well - one thing I'd want to make sure of - given where this runs in the pipeline (inline during the construction) - is that if a dynamic shape dimension is passed in as an argument and not constant at this point this is still useful. If it's not, it may be better to run this after outlining on the flow.executable where the dynamic dimensions have been materialized and passes like RematerializeDispatchConstants (which may inline some of those dynamic dimensions into the region after other passes run) have a chance to run. Today with where this is running if a dynamic dimension gets turned into a constant this may behave differently -- or may not - would be useful to test :)

Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: Thanks for pointing that out. This transformation won't help much if the dispatch sizes are none-constant. The reason I ran it at this stage is because otherwise the `flow.dispatch.input.load` ops don't seem to get transformed to static shape ops but I think we should be able to canonicalize later one.
I'll make those transformations run separately later on then.

Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: I moved the pass after `RematerializeDispatchConstants` but I'm not sure this is the right spot. After outlining I cannot easily access the dispatch count anymore right? I'm probably missing something.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+  AffineExpr step = getAffineSymbolExpr(symbols.size(), ctx);
+  symbols.push_back(stepVal);
+  LLVM_DEBUG(llvm::dbgs() << "Before: " << expr << "\n");
+  AffineExpr ee = substWithMin(expr, dimExpr, lb,
+                               lb + step * ((ub - 1) - lb).floorDiv(step));
+  LLVM_DEBUG(llvm::dbgs() << "After: " << ee << "\n");
+  return ee;
+}
+
+/// Traverse the `dims` and substitute known min or max expressions in place of
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:96]: Its very hard to follow what this method is supposed to do based on the comments.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::ForOp op,
+                                PatternRewriter &rewriter) const override {
+    bool alwaysRunFirstIteration = false;
+    bool neverRunSecondIteration = false;
+    MLIRContext *ctx = op->getContext();
+    // Calculate the minimum value of ub - lb. If it is strictly positive it
+    // means the loop will always run at least once.
+    {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:273]: Might just be easier to move this into a method `alwaysRunsFirstIteration`, similarly `neverRunSecondIteration` below. Then this method is just
```
if (!alwaysRunsFirstIteration(...) || !neverRunSecondIteration(...)) return failure();
```
and the rest of the code.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        for (unsigned dimIdx = 0; dimIdx < dims.size(); ++dimIdx) {
+          Value dim = dims[dimIdx];
+          AffineExpr dimExpr = getAffineDimExpr(dimIdx, expr.getContext());
+          LLVM_DEBUG(llvm::dbgs()
+                     << "Subst: " << dim << " @ " << dimExpr << "\n");
+          AffineExpr substitutedExpr;
+          if (auto forOp = scf::getForInductionVarOwner(dim))
+            substitutedExpr = substituteLoopInExpr(
+                expr, dimExpr, forOp.lowerBound(), forOp.upperBound(),
+                forOp.step(), dims, symbols);
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:117]: Add `continue` for each `if` to avoid unnecessary checks.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+namespace {
+/// Traverse `e` and return an AffineExpr where all occurrences of `dim` have
+/// been replaced by either:
+///  - `min` if `positivePath` is true when we reach an occurrence of `dim`
+///  - `max` if `positivePath` is true when we reach an occurrence of `dim`
+/// `positivePath` is negated each time we hit a multiplicative or divisive
+/// binary op with a constant negative coefficient.
+static AffineExpr substWithMin(AffineExpr e, AffineExpr dim, AffineExpr min,
+                               AffineExpr max, bool positivePath = true) {
+  if (e == dim) return positivePath ? min : max;
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:48]: Maybe a bit more readable if there is a `getMinOfExpr` and `getMaxOfExpr`. Those are useful utilities to  have in general I think. I have encountered this a couple of times and was going to add this upstream (but I in the end didnt really needed it so didnt go through with it)

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:48]: Actually maybe this comment is wrong. I see how `positivePath` makes this more involved. But nevertheless leaving the comment here in case this make sense to you

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+  if (auto bin = e.dyn_cast<AffineBinaryOpExpr>()) {
+    AffineExpr lhs = bin.getLHS();
+    AffineExpr rhs = bin.getRHS();
+    if (bin.getKind() == mlir::AffineExprKind::Add)
+      return substWithMin(lhs, dim, min, max, positivePath) +
+             substWithMin(rhs, dim, min, max, positivePath);
+
+    auto c1 = bin.getLHS().dyn_cast<AffineConstantExpr>();
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:56]: The below holds only for `AffineExprKind::Mul`? What about `FloorDiv` and `CeilDiv`? I would assume that when `positivePath=true` the denominator must be substituted with the max value.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+  return ee;
+}
+
+/// Traverse the `dims` and substitute known min or max expressions in place of
+/// induction variables in `exprs`.
+static AffineMap substitute(AffineMap map, SmallVectorImpl<Value> &dims,
+                            SmallVectorImpl<Value> &symbols) {
+  auto exprs = llvm::to_vector<4>(map.getResults());
+  bool simplified;
+  do {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:102]: Is this `do` loop really needed? Once a `map` is simplified do we need to simplify it again?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+/// `linalg::AffineMinSCFCanonicalizationPattern`. In the future we could unify
+/// this code back by upstreaminng an interface to MLIR to represent generic ID
+/// ops.
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+/// Traverse `e` and return an AffineExpr where all occurrences of `dim` have
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:40]: Agree, this can be moved to core however it will require some kind of interfaceOp to be able to interact with the IREE specific ops. I don't mind doing this right now but I'm not convince at this point I have something stable enough. 

The way to have this in core in my opinion is to have an InterfaceOp "IdRange" (or a better name) allowing to query the range of a given op.
Do you see another solution to have this in core? Do you think this should be done at this point or do you expect that further change in IREE flow may affect what we are doing there.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+/// been replaced by either:
+///  - `min` if `positivePath` is true when we reach an occurrence of `dim`
+///  - `max` if `positivePath` is true when we reach an occurrence of `dim`
+/// `positivePath` is negated each time we hit a multiplicative or divisive
+/// binary op with a constant negative coefficient.
+static AffineExpr substWithMin(AffineExpr e, AffineExpr dim, AffineExpr min,
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:46]: This function is copied from core (AffineMinSCFCanonicalizationPattern as point in the comment) so I would rather not modify it as the plan is to merge it back.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 6 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:56]: Below holds for Mul, Mod, FloorDiv and CeilDiv (everything but Add). Yes if the denominator is a negative constant we need to swap the `positivePath` value. This is what is being done.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:71]: Updated the comment and removed some of the code to handle affin_min within ForOp that we don't need at the moment to make things simpler.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp]: Actually we do want the end of the loop to be executed. Removed some of this code and replaced the rest by `else if`

**Comment by ThomasRaoux** [comment]
I simplified the PR a bit by removing some case that are currently not needed. Please take another look.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 6 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:56]: As I understand what is happening here, you want to take an expression and find its "minimum" based on the dim being between a certain range right?

so if `d0` is in range `[1, 50]` and you have `e / d0` wont min of this expression be if you replace `d0` with its `max` value ? i.e. `e / 50`? Maybe I am missing the intent

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:56]: Hmmm, as I write that, i realize that `e/d0` cant happen. Its not affine. So its OK. Ill take a look again

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:40]: If you take this `dim` here, there is nothing that says this comes from an ID op. It is only working under the constraint that `dim` is in `[min, max)`. I am not saying move the entire pattern to core. Just this method as a utility. If it is in core already, just move it to a convenient place and expose it as a utility method. There are a bunch of such methods in MLIR.

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir (hunk 1 above)
Comment (benvanik) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: > This transformation won't help much if the dispatch sizes are none-constant.

ah ok thanks for explaining! That would be useful to note in the pass description ("...when the dispatch count is constant"). Renaming the pass would also help indicate when it's used more clearly: `PropagateConstantDispatchCounts` or something like it (following the pattern of "verb adjective noun" to make it clear what's being done to what and when, and to someone looking at the pass list make more sense as to when it should run). The more general 'canonicalization' name makes me think this is doing more than that and more generally useful.

I'm interested to know how load-bearing this is: is this strictly IR cleanup to make things easier to read (which is totally fine!) or does this have a significant impact? Anything making assumptions about the dispatch count is at the flow level generally an anti-pattern when dynamic shapes are involved. It may even be invalid for this kind of optimization even in static cases: workgroup sizes are not yet decided as until we go into the HAL we intentionally are target backend agnostic, meaning that a workgroup count of x=50 here in flow may end up being 2 in hal if the workgroup size for a particular backend is chosen to be 25, etc.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:46]: If the function is copied from core, even more reason to just expose this as a utility method and use it here. If it is good enough for IREE, it is good enough for core. If there are fixes to this upstream we will then have to manually port that here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

**Review by benvanik — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:40]: I see, I didn't realize this was an option as it seems like exposing helper functions but in this case it does seem reasonable.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:46]: I see, I didn't realize it was an option to just expose it. I can re-use this one function. I don't think any of the rest of the code can go to core on their own though.

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir (hunk 1 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: I agree, I missed the part of where it is added to the pipeline. I would actually not make any decisions here at Flow level w.r.t to number of workgroups, etc. When lowering from Flow to HAL you run the codegen passes on the computation. It should be done there....

Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: > is this strictly IR cleanup to make things easier to read (which is totally fine!) or does this have a significant impact?

It is an optimization to remove useless control flow when the dispatch size is static and this also allows vectorization to work as it currently require static shape.

>I agree, I missed the part of where it is added to the pipeline. I would actually not make any decisions here at Flow level w.r.t to number of workgroups, etc. When lowering from Flow to HAL you run the codegen passes on the computation. It should be done there....

Mahesh, so you are saying this is not using the right dialect? What Ops should it be based on?

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: If I am not mistaken there are corresponding operations in HAL dialect as well. Ideally I would trigger this after we have `LaunchConfig`. Because that is determining the static tile size to use, and the workgroup size. Then you also have the number of workgroups. That is where I was thinking this could slot in. Am I missing something?

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:40]: Sent a patch to move this into a helper in core:
https://reviews.llvm.org/D94826

File: iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: I see so I should be using `hal.interface.workgroup.id` instead? Can I still access the number of workgroup dispatched in this case? It seemed like it got moved into the function caller?

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: Maybe you are right about that. I have to still get upto speed on that (backlog flushing some failures on patches that are ready to land). Are you planning to do vectorization at flow level. That would be problematic in terms of making things work with everything else that already exists.

Comment (nicolasvasilache) [reply iree/compiler/Dialect/Flow/Transforms/test/dispatch_id_canonicalization.mlir:19]: We should def. aim at making this work with non-constant quantities in the future.
e.g. `for i = 0; i < ub; i += workgroup_size` where we also know that `ub == cst * workgroup_size` or even just `ub % workgroup_size == 0`.
No need to make logic more complex now but let's add a TODO please.

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
Let's please kill all the copy-pasta.
What this does is injecting more cases into the guts of the core substitute function and this can be done with lambdas.

File: iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+struct AffineMinSCFCanonicalizationPattern
+    : public OpRewritePattern<AffineMinOp> {
+  using OpRewritePattern<AffineMinOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineMinOp minOp,
+                                PatternRewriter &rewriter) const override;
+};
+
+LogicalResult AffineMinSCFCanonicalizationPattern::matchAndRewrite(
```
[the remark this answers is not in the document]
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp:159]: I'm sent a patch to do that (https://reviews.llvm.org/D94998). However if we want to support removing the scf.forOp I need to access `substitute` in this pattern. Do you think `substitute` needs to be exposed too?
